File: StartTimerDialogController.m
//
//  StartTimerDialogController.m
//  MenuTimer
//
//  Created by Kristopher Johnson on 3/29/09.
//  Copyright 2009 Capable Hands Technologies, Inc.. All rights reserved.
//
//  This file is part of Menubar Countdown.
//
//  Menubar Countdown is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  Menubar Countdown is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with Menubar Countdown.  If not, see <http://www.gnu.org/licenses/>.
//

#import "StartTimerDialogController.h"
#import "UserDefaults.h"


@implementation StartTimerDialogController

- (void)showDialog {
    [NSApp activateIgnoringOtherApps:YES];
    if (![startTimerDialog isVisible]) {
        [startTimerDialog center];
        [startTimerDialog makeFirstResponder:nil];
    }
    [startTimerDialog makeKeyAndOrderFront:self];
}


- (IBAction)dismissDialog:(id)sender {
    if (![startTimerDialog makeFirstResponder:nil]) {
        // TODO: Figure out what to do if responder didn't resign
    }
    [startTimerDialog orderOut:sender];
}


- (NSTimeInterval)timerInterval {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    int timerHours = [defaults integerForKey:UserDefaultsTimerHoursKey];
    int timerMinutes = [defaults integerForKey:UserDefaultsTimerMinutesKey];
    int timerSeconds = [defaults integerForKey:UserDefaultsTimerSecondsKey];
    return (timerHours * 3600) + (timerMinutes * 60) + timerSeconds;
}


- (void)dealloc {
    [startTimerDialog release];
    [super dealloc];
}


@end

File: MenubarCountdown/StartTimerDialogController.swift
//
//  StartTimerDialogController.swift
//  MenubarCountdown
//
//  Copyright © 2009,2015,2019 Kristopher Johnson. All rights reserved.

import Cocoa

/**
 Window controller for the "Menubar Countdown Settings" dialog that is used to start the countdown.

 Also see StartTimerDialog.xib.
 */
final class StartTimerDialogController: NSWindowController {
    @IBOutlet var startTimerDialog: NSWindow!

    /**
     Load the dialog from the nib.

     The `owner` argument must be an object with a `startTimerDialogController`
     property, which will be set to an instance of this class.
     */
    static func load(owner: NSObject) {
        Bundle.main.loadNibNamed("StartTimerDialog",
            owner: owner,
            topLevelObjects: nil)
    }

    /**
     Display the dialog and bring it to the front.
     */
    func showDialog() {
        NSApp.activate(ignoringOtherApps: true)
        if !startTimerDialog.isVisible {
            startTimerDialog.center()
            startTimerDialog.makeFirstResponder(nil)
        }
        startTimerDialog.makeKeyAndOrderFront(self)
    }

    /**
     Hide the dialog.
     */
    @IBAction func dismissDialog(_ sender: AnyObject) {
        if !startTimerDialog.makeFirstResponder(nil) {
            Log.error("first responder didn't resign")
        }
        startTimerDialog.orderOut(sender);
    }
}
